### ISSUE TITLE
Sigmoid neurons don't have intercepts

### ISSUE THREAD

**Issue by arvoelke** [nengo/nengo#1211]
```
import nengo
import pylab

with nengo.Network() as model:
    x = nengo.Ensemble(4, 1, neuron_type=nengo.neurons.Sigmoid(),
                       intercepts=[-1, -0.5, 0, 0.5])

with nengo.Simulator(model) as sim: pass

u = np.linspace(-1, 1, 100000)
acts = nengo.builder.ensemble.get_activities(sim.data, x, u[:, None])
assert ((acts > 0).all())

pylab.figure()
pylab.plot(u[:, None] * sim.data[x].encoders.T, acts)
pylab.show()
```

![download](https://cloud.githubusercontent.com/assets/5420057/20606344/494d660c-b23e-11e6-9a20-9313fcc7fbe8.png)

It seems that the desired intercept is achieved by shifting the curves down by approximately 50. I think the effect of choosing intercepts for this model should be documented (or fixed if this is incorrect).

**Comment by tbekolay** [comment]
Soooo yeah... Sigmoid neurons were asked for as a 2.0 feature and no one else volunteered to implement them, so I did my best. There's some explanation of how I interpreted the intercept in the original PR (#570) and in [this tech report that I never finished](https://github.com/ctn-archive/technical-reports/blob/sigmoid_in_nengo/sigmoid.ipynb).

What do you think would be the correct way to interpret the intercept? If we have the sigmoid actually go down to 0 by the intercept, the sigmoids are going to be pretty close to step functions a lot of the time. Maybe that's better for some applications, but I did it this way to have the most varied activity across the radius of the ensemble.

**Comment by arvoelke** [comment]
Ah, so `f(midpoint(1, intercepts)) = 0.5 * max_rates` from the other thread. I see.

An alternative would be to specify some firing rate `b` when you construct the sigmoid such that `f(intercept) = b`. The way you've done it now seems roughly equivalent to setting `b = 50`, and might be less arbitrary?

Another alternative would be shift the curves down by `b`, so that you actually get the requested intercept. Not sure what you about this being close to step functions. It would be close to a rectified quadratic?

I'm also just happy with documenting the first formula so that people have a way of understanding what happened to their intercepts.

**Comment by tbekolay** [comment]
> An alternative would be to specify some firing rate b when you construct the sigmoid such that f(intercept) = b

Yeah, I'd be down with that. Most people would use the default but being able to change it is nice :+1:

> Another alternative would be shift the curves down by b, so that you actually get the requested intercept.

Since I consider the output of these neuron types to be firing rates, I assumed that a negative firing rate was invalid. But it's true, negative firing rates could be a thing, would be interested to know what that would mean for the decoding process etc.

> I'm also just happy with documenting the first formula so that people have a way of understanding what happened to their intercepts.

Yep, agreed. Probably makes sense to make a notebook for it, like we do for the Izhikevich neurons. I'll put that on my todo list, unless you're keen to make the PR (either way's good with me).

**Comment by arvoelke** [comment]
>  I assumed that a negative firing rate was invalid. But it's true, negative firing rates could be a thing, would be interested to know what that would mean for the decoding process etc.

I forgot to mention rectification. I've gone back and edited my post.  But this may also be distinctly different from the sigmoids that people expect.

That said, negative firing rates are fine for neuron models in Nengo. I've tried it out before and it is handled well by decoding / etc. I haven't tried a spiking version, but it would probably be fine as well.

Maybe we should have @hunse weigh in on the first suggestion of picking a `b` such that `f(intercept) = b` since he is the one mainly using this model? There might be a better free parameter to play with.

**Comment by arvoelke** [comment]
Some additional context on the negative firing rate in case anyone's curious: echo-state networks use tanh rate neurons (similar to our sigmoid) that are shifted (but not rectified) so that they effectively fire with negative "rates". This can be reproduced in Nengo, and is often super accurate for rate-mode simulations (since we often don't want that sharp transition that rectification gives you in the decoded function).

**Comment by hunse** [comment]
So the current formulation basically says ``intercepts == biases``, i.e. the inflection point of the sigmoid is at the intercept. This is easy for me in the types of models I deal with (machine learning ones), and will probably be intuitive for others with an ML background (and ML-type models are where I usually expect this to be used). So I'd like to keep the default the way it is now.

Having more control over this could be nice. The ``f(intercept) = b`` formulation makes sense. I would specify ``b`` in terms of the pure [0, 1] sigmoid function, though. So that means currently, ``b = 0.5``. Another reasonable choice would be ``b = sigmoid(-sigmoid(0) / d_sigmoid(0)) = sigmoid(-2) = 0.1192``, since this would be the intercept if you drew a straight line down from the inflection point along the tangent there (where ``d_sigmoid`` is the derivative). Letting the user select any value for ``b`` makes sense to me.

I'm not so sure about rectifying or allowing negative rates. Neither of these seem to me to be something I would use. But if it'd be useful for someone, we could probably figure out a way to add another parameter to control that.

**Comment by arvoelke** [comment]
> Having more control over this could be nice. The f(intercept) = b formulation makes sense. I would specify b in terms of the pure [0, 1] sigmoid function, though. So that means currently, b = 0.5.

Oh that makes even more sense then. :) Since I don't have a need to play with this parameter though, I can make a documentation change PR that just compactly states what you said.

**Comment by Seanny123** [comment]
Resolved in #1227
